File: src/javascript/algorithms/tree/10.kthSmallestBST/kthSmallestBST.js
//Iterative Inorder BST: TC: O(n) SC: O(n)
function kthSmallestBST1(root, k) {
    let stack = [];
    let currNode = root;
    while(currNode || stack.length >0) {
        while(currNode) {
            stack.push(currNode);
            currNode = currNode.left;
        }
        currNode = stack.pop();
        k--;
        if(k === 0) return currNode.value;
        currNode = currNode.right;
    }

    return 0;
}

//Recursive Inorder BST: TC: O(n) SC: O(n)
function kthSmallestBST2(root, k) {
    let list = [];
    inorderTraversal(root, list);
    return list[k-1];
}

function inorderTraversal(node, list){
    if(node === null) return;
    inorderTraversal(node.left, list);
    list.push(node.value);
    inorderTraversal(node.right, list);
}

class TreeNode {
    constructor(value) {
      this.left = null;
      this.right = null;
      this.value = value;
    }
}


let root1 = new TreeNode(5);
root1.left = new TreeNode(1);
root1.right = new TreeNode(8);
root1.left.left = new TreeNode(0);
root1.left.right = new TreeNode(2);
root1.right.left = new TreeNode(7);
root1.right.right = new TreeNode(9);

let root2 = new TreeNode(2);
root2.left = new TreeNode(1);
root2.right = new TreeNode(3);

console.log(kthSmallestBST1(root1, 3));
console.log(kthSmallestBST2(root1, 3));

console.log(kthSmallestBST1(root2, 2));
console.log(kthSmallestBST2(root2, 2));

File: src/python/algorithms/tree/10.kthSmallestBST/kth_smallest_bst.py
"""
Kth Smallest Element in a BST

Given the root of a binary search tree and an integer k,
return the kth smallest value (1-indexed) of all values in the BST.

Time Complexity: O(n)
Space Complexity: O(n)
"""


class TreeNode:
    def __init__(self, value=0):
        self.value = value
        self.left = None
        self.right = None


def kth_smallest_iterative(root, k):
    """
    Iterative inorder traversal using stack.
    TC: O(n), SC: O(n)
    """
    if not root or k <= 0:
        return None
    
    stack = []
    curr = root
    
    while curr or stack:
        while curr:
            stack.append(curr)
            curr = curr.left
        
        curr = stack.pop()
        k -= 1
        
        if k == 0:
            return curr.value
        
        curr = curr.right
    
    return None


def kth_smallest_recursive(root, k):
    """
    Recursive inorder traversal.
    TC: O(n), SC: O(n)
    """
    if not root or k <= 0:
        return None
    
    result = []
    
    def inorder(node):
        if not node:
            return
        inorder(node.left)
        result.append(node.value)
        inorder(node.right)
    
    inorder(root)
    return result[k - 1] if k <= len(result) else None


# Test cases
if __name__ == "__main__":
    # Create BST: [5, 1, 8, 0, 2, 7, 9]
    root1 = TreeNode(5)
    root1.left = TreeNode(1)
    root1.right = TreeNode(8)
    root1.left.left = TreeNode(0)
    root1.left.right = TreeNode(2)
    root1.right.left = TreeNode(7)
    root1.right.right = TreeNode(9)
    
    print(f"3rd smallest (iterative): {kth_smallest_iterative(root1, 3)}")  # 2
    print(f"3rd smallest (recursive): {kth_smallest_recursive(root1, 3)}")  # 2
    
    # Create BST: [2, 1, 3]
    root2 = TreeNode(2)
    root2.left = TreeNode(1)
    root2.right = TreeNode(3)
    
    print(f"2nd smallest (iterative): {kth_smallest_iterative(root2, 2)}")  # 2
    print(f"2nd smallest (recursive): {kth_smallest_recursive(root2, 2)}")  # 2
    print(f"10th smallest (out of range): {kth_smallest_iterative(root2, 10)}")  # None
